td-shim/src/bin/td-shim/info.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+
+use alloc::vec::Vec;
+use core::mem::size_of;
+use core::str::FromStr;
+use log::error;
+use scroll::{Pread, Pwrite};
+use td_layout::build_time::{TD_SHIM_FIRMWARE_BASE, TD_SHIM_FIRMWARE_SIZE};
+use td_layout::memslice;
+use td_shim::metadata::*;
+use td_shim::speculation_barrier;
+use td_shim::{PayloadInfo, TdKernelInfoHobType, TD_ACPI_TABLE_HOB_GUID, TD_KERNEL_INFO_HOB_GUID};
+use td_uefi_pi::pi::guid::Guid;
+use td_uefi_pi::pi::hob::*;
+use td_uefi_pi::{fv, hob, pi};
+
+pub struct ShimInfo<'a> {
+    // If metadata contains the `TD_HOB` section, TD-Shim
+    // can get additional information from TD HOB
+    pub td_hob: Option<&'static [u8]>,
+
+    // If metadata contains one/more `PermMem` sections,
+    // TD-Shim should ignore the memory information in TD HOB.
+    metadata_has_perm: bool,
+
+    pub memory: Vec<ResourceDescription>,
+    pub acpi_tables: Vec<&'a [u8]>,
+    pub payload_info: Option<PayloadInfo>,
+    pub payload_param: Option<&'a [u8]>,
+}
+
+impl<'a> ShimInfo<'a> {
+    // Validate the metadata and get the basic infomation from
+    // it if any
+    pub fn new() -> Option<Self> {
+        let metadata_offset = unsafe { *((u32::MAX - TDX_METADATA_OFFSET + 1) as *const u32) };
+        if metadata_offset >= TD_SHIM_FIRMWARE_SIZE
+            || metadata_offset < size_of::<Guid>() as u32
+            || metadata_offset > TD_SHIM_FIRMWARE_SIZE - size_of::<TdxMetadataDescriptor>() as u32
+        {
+            error!("Invalid TDX Metadata offset\n");
```

Comment:
I remember we support both OFFSET style and GUID style. Should we support both cases here?

Reply:
There is always an `OFFSET` at `4GiB - 0x20` no matter if there is a GUID chain or not